From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ratchanan Srirattanamet <ratchanan@ubports.com>
Date: Thu, 23 Oct 2025 02:43:55 +0700
Subject: [PATCH] card: notify audio HAL about headphone/headset connection
 state

Audio HAL implementation on MediaTek Android 12+ devices
tracks headphone/headset connection state internally and will not
write buffer to underlying audio device if it considers the port
disconnected.

Use AUDIO_PARAMETER_DEVICE_CONNECT/AUDIO_PARAMETER_DEVICE_DISCONNECT
to notify it about current headphone/headset connection state.

This implementation builds upon Nikita (NotKit, TheKit)'s implementation
and moves the call to `pa_droid_set_parameters()` to be inside `module-
droid-card.c` itself. This removes duplicated code and separates the
notification code from detection code. It also makes device id lookup
more dynamic reusing existing table.

It's not clear whether SailfishOS uses PulseAudio port availability to
control which port to direct audio to, so it's not clear whether this
patch is applicable for upstream or not.

Co-authored-by: TheKit <nekit1000@gmail.com>
Origin: vendor
Forwarded: No
Bug-UBports: https://gitlab.com/ubports/porting/reference-device-ports/halium12/volla-x23/volla-vidofnir/-/issues/7
Bug-UBports: https://gitlab.com/ubports/development/ubuntu-touch/-/issues/2280
Signed-off-by: Herman van Hazendonk <github.com@herrie.org>
Upstream-Status: Submitted [https://github.com/droidian/pulseaudio-modules-droid-modern/pull/2]

---
 src/common/conversion.c               |  4 +++
 src/common/include/droid/conversion.h |  1 +
 src/droid/module-droid-card.c         | 48 +++++++++++++++++++++++++++
 3 files changed, 53 insertions(+)

diff --git a/src/common/conversion.c b/src/common/conversion.c
index b67017b..9af7978 100644
--- a/src/common/conversion.c
+++ b/src/common/conversion.c
@@ -217,6 +217,10 @@ bool pa_droid_output_port_name(audio_devices_t value, const char **to_str) {
     return string_convert_num_to_str(string_conversion_table_output_device_fancy, (uint32_t) value, to_str);
 }
 
+bool pa_droid_output_port_name_to_device(const char *str, audio_devices_t *to_value) {
+    return string_convert_str_to_num(string_conversion_table_output_device_fancy, str, to_value);
+}
+
 bool pa_droid_input_port_name(audio_devices_t value, const char **to_str) {
     return string_convert_num_to_str(string_conversion_table_input_device_fancy, (uint32_t) value, to_str);
 }
diff --git a/src/common/include/droid/conversion.h b/src/common/include/droid/conversion.h
index 2d134a8..23caed4 100644
--- a/src/common/include/droid/conversion.h
+++ b/src/common/include/droid/conversion.h
@@ -79,6 +79,7 @@ bool pa_input_device_default_audio_source(audio_devices_t input_device, audio_so
 
 /* Pretty port names */
 bool pa_droid_output_port_name(audio_devices_t value, const char **to_str);
+bool pa_droid_output_port_name_to_device(const char *str, audio_devices_t *to_value);
 bool pa_droid_input_port_name(audio_devices_t value, const char **to_str);
 
 int pa_conversion_parse_list(pa_conversion_string_t type, const char *separator,
diff --git a/src/droid/module-droid-card.c b/src/droid/module-droid-card.c
index beb81cc..38b8926 100644
--- a/src/droid/module-droid-card.c
+++ b/src/droid/module-droid-card.c
@@ -23,6 +23,7 @@
 #include <config.h>
 #endif
 
+#include <inttypes.h>
 #include <signal.h>
 #include <stdio.h>
 
@@ -59,6 +60,7 @@
 //#include <droid/hardware/audio_policy.h>
 //#include <droid/system/audio_policy.h>
 
+#include <droid/conversion.h>
 #include <droid/droid-util.h>
 #include <droid/sllist.h>
 #include <droid/utils.h>
@@ -784,6 +786,47 @@ static int card_set_profile(pa_card *c, pa_card_profile *new_profile) {
     return 0;
 }
 
+static pa_hook_result_t port_availability_changed_hook_callback(void *hook_data,
+                                                                void *call_data,
+                                                                void *slot_data) {
+    pa_device_port *port = call_data;
+    struct userdata *u = slot_data;
+
+    /* Not meant for us */
+    if (port->card != u->card)
+        return PA_HOOK_OK;
+
+    /* Track only outputs for now. */
+    if (port->direction != PA_DIRECTION_OUTPUT)
+        return PA_HOOK_OK;
+
+    /* We don't track availability of this port. */
+    if (port->available == PA_AVAILABLE_UNKNOWN)
+        return PA_HOOK_OK;
+
+    const char * verb = port->available == PA_AVAILABLE_YES
+        ? AUDIO_PARAMETER_DEVICE_CONNECT
+        : AUDIO_PARAMETER_DEVICE_DISCONNECT;
+
+    uint32_t device;
+    if (!pa_droid_output_port_name_to_device(port->name, &device)) {
+        pa_log_warn("Can't notify Android of port '%s' as it's unknown.",
+            port->name);
+        return PA_HOOK_OK;
+    }
+
+    pa_log_info("Notifying Android of port '%s' (%" PRIu32 ") becoming %s.",
+        port->name, device,
+        port->available == PA_AVAILABLE_YES ? "available" : "not available");
+
+    char * setparam = pa_sprintf_malloc("%s=%" PRIu32 "", verb, device);
+    pa_droid_set_parameters(u->hw_module, setparam);
+
+    pa_xfree(setparam);
+
+    return PA_HOOK_OK;
+}
+
 
 int pa__init(pa_module *m) {
     struct userdata *u = NULL;
@@ -908,6 +951,11 @@ int pa__init(pa_module *m) {
     else
         u->extevdev = NULL;
 
+    pa_module_hook_connect(u->module,
+                           &u->module->core->hooks[PA_CORE_HOOK_PORT_AVAILABLE_CHANGED],
+                           PA_HOOK_NORMAL,
+                           port_availability_changed_hook_callback, u);
+
     pa_card_put(u->card);
 
     return 0;
-- 
2.53.0

